File: ANSv4/Catalogo3/classes/Pedido.vb
Option Explicit On

Public Class Pedido

    ' Define como se llama este modulo para el control de errores
    Private Const m_sMODULENAME_ As String = "clsPedido"

    Private Conexion1 As System.Data.OleDb.OleDbConnection  ' diego    Private Conexion1 As OleDbConnection

    Private mvarNroPedido As String 'copia local
    Private mvarF_Pedido As Date 'copia local
    Private mvarIdCliente As Long 'copia local
    Private mvarNroImpresion As Byte 'copia local
    Private mvarObservaciones As String 'copia local
    Private mvarTransporte As String 'copia local

    Private DetallePedido As Collection

    Public Event GuardoOK(ByVal NroPedido As String)

    Public Property Transporte() As String
        Set(ByVal value As String)
            mvarTransporte = value
        End Set

        Get
            Return mvarTransporte
        End Get
    End Property

    Public Property Observaciones() As String
        Set(ByVal value As String)
            mvarObservaciones = value
        End Set
        Get
            Return mvarObservaciones
        End Get
    End Property

    Public ReadOnly Property CantidadItems() As Long
        Get
            Return DetallePedido.Count
        End Get
    End Property

    Public Sub Nuevo(Optional ByVal IDdeCliente As Long = 0)
        DetallePedido = Nothing
        DetallePedido = New Collection
        mvarNroPedido = 0
        mvarF_Pedido = Date.Today
        If IDdeCliente > 0 Then mvarIdCliente = IDdeCliente
    End Sub

    ' FUNDAMENTAL PARA QUE TE DE LOS NOMBRES
    Public Function Renglon(ByVal Numero As Integer) As PedidoItem
        Return DetallePedido("_" & Numero)
    End Function

    Public WriteOnly Property Conexion() As System.Data.OleDb.OleDbConnection
        Set(ByVal value As System.Data.OleDb.OleDbConnection)
            Conexion1 = value
        End Set
    End Property

    Public Property NroImpresion() As Byte
        Set(ByVal value As Byte)
            mvarNroImpresion = value
        End Set
        Get
            Return mvarNroImpresion
        End Get
    End Property

    Public ReadOnly Property IdCliente() As Long
        Get
            Return mvarIdCliente
        End Get
    End Property

    Public ReadOnly Property F_Pedido() As Date
        Get
            Return mvarF_Pedido
        End Get
    End Property

    Public ReadOnly Property NroPedido() As String
        Get
            Return mvarNroPedido
        End Get
    End Property

    'diego     Private Sub Class_Terminate()
    'diego         DetallePedido = Nothing
    'diego     End Sub

    Private Function ValidarConexion() As Boolean
        Return Not (Conexion1 Is Nothing)
    End Function

    Public Sub Guardar(Optional ByVal Origen As String = vbNullString)

        If Not (ValidarConexion()) Then Exit Sub

        On Error GoTo ErrorHandler

        Dim rec As System.Data.OleDb.OleDbDataReader
        Dim cmd As New System.Data.OleDb.OleDbCommand

        If DetallePedido.Count < 1 Then Exit Sub

        If UCase(Origen) = "VER" Then
            Funciones.adoModulo.adoComandoIU(Conexion1, "DELETE FROM tblPedido_Enc WHERE NroPedido='09999-99999999'")
            mvarNroPedido = "09999-99999999"
        Else

            If Len(Trim(vg.NroPDR)) > 0 Then
                Funciones.adoModulo.adoComandoIU(Conexion1, "DELETE FROM tblPedido_Enc WHERE NroPedido='" & vg.NroPDR & "'")
                vg.NroPDR = ""
            End If

            rec = Funciones.adoModulo.adoComando(Conexion1, "SELECT TOP 1 right(NroPedido,8) AS NroPedido FROM tblPedido_Enc WHERE left(NroPedido,5)=" & vg.NroUsuario & " ORDER BY NroPedido DESC")
            If Not rec.HasRows Then
                If vg.miSABOR = 2 Then ' ES UN CLIENTE
                    mvarNroPedido = Trim(vg.NroUsuario) & "-C0000001"
                Else
                    mvarNroPedido = Trim(vg.NroUsuario) & "-00000001"
                End If
            Else
                If vg.miSABOR = 2 Then ' ES UN CLIENTE
                    mvarNroPedido = Trim(vg.NroUsuario) & "-C" & Format(CLng(Right(rec("NroPedido"), 7)) + 1, "0000000")
                Else
                    mvarNroPedido = Trim(vg.NroUsuario) & "-" & Format(CLng(Right(rec("NroPedido"), 7)) + 1, "00000000")
                End If
            End If
            rec = Nothing

        End If

        cmd.Parameters.Add("pNroPedido", SqlDbType.VarChar, 14).Value = mvarNroPedido
        cmd.Parameters.Add("pF_Pedido", SqlDbType.Date).Value = mvarF_Pedido
        cmd.Parameters.Add("pIdCliente", SqlDbType.Int).Value = mvarIdCliente
        cmd.Parameters.Add("pNroImpresion", SqlDbType.Int).Value = 0
        cmd.Parameters.Add("pObservaciones", SqlDbType.VarChar, 128).Value = mvarObservaciones
        cmd.Parameters.Add("pTransporte", SqlDbType.VarChar, 128).Value = mvarTransporte

        cmd.Connection = Conexion1
        cmd.CommandType = CommandType.StoredProcedure
        cmd.CommandText = "usp_Pedido_Enc_add"
        cmd.ExecuteNonQuery()

        cmd = Nothing

        ' Ahora GUARDO los ITEMS de ESTE pedido

        Dim I As Integer
        Dim wTotal As Double
        wTotal = 0

        For I = 1 To DetallePedido.Count
            GuardarItem(Renglon(I).IDCatalogo, _
                        Renglon(I).cantidad, _
                        Renglon(I).Similar, _
                        Renglon(I).Bahia, _
                        Renglon(I).Oferta, _
                        Renglon(I).Deposito, _
                        Renglon(I).Precio, _
                        Renglon(I).Observaciones)
            wTotal = wTotal + (Renglon(I).cantidad * Renglon(I).Precio)
        Next I

        vg.NroImprimir = mvarNroPedido

        ' Por POLITICA NUESTRA  -- >  SE BORRA
        If UCase(Origen) = "VER" Then
            ' Actualizo Fecha de Transmicion para que no lo mande
            'diego             adoModulo.adoComandoIU(vg.Conexion, "EXEC usp_Pedido_Transmicion_Upd '" & mvarNroPedido & "'")
        Else
            vg.auditor.Guardar(ObjetosAuditados.Pedido, AccionesAuditadas.EXITOSO, "cli:" & Format(mvarIdCliente, "000000") & " ped:" & mvarNroPedido & " tot:" & Format(wTotal, "fixed"))
            Nuevo()
            RaiseEvent GuardoOK(vg.NroImprimir)
        End If

        Exit Sub

ErrorHandler:

        'diego         adoREC = Nothing
        cmd = Nothing
        Err.Raise(Err.Number, Err.Source, Err.Description)

    End Sub

    Public Sub ADDItem(ByVal IDCatalogo As String, _
                       ByVal cantidad As Integer, _
                       ByVal Similar As Boolean, _
                       ByVal Bahia As Boolean, _
                       ByVal Oferta As Boolean, _
                       ByVal Deposito As Byte, _
                       ByVal Precio As Single, _
                       ByVal Observaciones As String)

        'OJO chkEsOfertaBahia===Bahia

        Dim mvarItem As New PedidoItem

        mvarItem.IDCatalogo = IDCatalogo
        mvarItem.cantidad = cantidad
        mvarItem.Similar = Similar
        mvarItem.Bahia = Bahia
        mvarItem.Oferta = Oferta
        mvarItem.Deposito = Deposito
        mvarItem.Precio = Precio
        mvarItem.Observaciones = Observaciones

        DetallePedido.Add(mvarItem, "_" & DetallePedido.Count + 1)

        mvarItem = Nothing

    End Sub

    Private Sub GuardarItem(ByVal IDCatalogo As String, _
                            ByVal cantidad As Integer, _
                            ByVal Similar As Boolean, _
                            ByVal Bahia As Boolean, _
                            ByVal Oferta As Boolean, _
                            ByVal Deposito As Byte, _
                            ByVal Precio As Single, _
                            ByVal Observaciones As String)

        If Not (ValidarConexion()) Then Exit Sub

        On Error GoTo ErrorHandler

        Dim rec As System.Data.OleDb.OleDbDataReader
        Dim cmd As New System.Data.OleDb.OleDbCommand

        cmd.Parameters.Add("pNroPedido", SqlDbType.VarChar, 14).Value = mvarNroPedido
        cmd.Parameters.Add("pIdCatalogo", SqlDbType.VarChar, 38).Value = IDCatalogo
        cmd.Parameters.Add("pCantidad", SqlDbType.Int).Value = cantidad
        cmd.Parameters.Add("pSimilar", SqlDbType.Bit).Value = Similar
        cmd.Parameters.Add("pOferta", SqlDbType.Bit).Value = Oferta
        cmd.Parameters.Add("pBahia", SqlDbType.Bit).Value = Bahia
        cmd.Parameters.Add("pDeposito", SqlDbType.TinyInt).Value = Deposito
        cmd.Parameters.Add("pPrecio", SqlDbType.Float).Value = Precio
        cmd.Parameters.Add("pObservaciones", SqlDbType.VarChar, 80).Value = Observaciones

        cmd.Connection = Conexion1
        cmd.CommandType = CommandType.StoredProcedure
        cmd.CommandText = "usp_Pedido_Det_add"
        cmd.ExecuteNonQuery()

        rec = Funciones.adoModulo.adoComando(vg.Conexion, "SELECT ID FROM CatalogoBAK WHERE ID='" & IDCatalogo & "'")
        If Not rec.HasRows Then
            Funciones.adoModulo.adoComandoIU(vg.Conexion, "EXEC usp_AnexaItemCatalogoBak '" & IDCatalogo & "'")
        Else
            Funciones.adoModulo.adoComandoIU(vg.Conexion, "EXEC usp_CambiaCodigoCatalogoBak '" & IDCatalogo & "'")
        End If
        rec = Nothing
        cmd = Nothing
        Exit Sub

ErrorHandler:

        rec = Nothing
        cmd = Nothing
        Err.Raise(Err.Number, Err.Source, Err.Description)

    End Sub


End Class

File: ANSv4/Catalogo4/_pedidos/Pedido.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace Catalogo._devoluciones
{
    public class Pedido
    {
        private const string m_sMODULENAME_ = "clsPedido";
        private string mvarNroPedido;
        private System.DateTime mvarF_Pedido;
        private int mvarIdCliente;
        private byte mvarNroImpresion;
        private string mvarObservaciones;
        private string mvarTransporte;
        private System.Data.OleDb.OleDbConnection mvarConexion;
        private System.Collections.Generic.Dictionary<string, PedidoItem> DetallePedido;
        
        public event GuardoOKEventHandler GuardoOK;
        public delegate void GuardoOKEventHandler(string NroPedido);

        public string Transporte
        {
            get { return mvarTransporte; }
            set { mvarTransporte = value; }
        }

        public string Observaciones
        {
            get { return mvarObservaciones; }
            set { mvarObservaciones = value; }
        }

        public int CantidadItems
        {
            get { return DetallePedido.Count; }
        }

       public Pedido(System.Data.OleDb.OleDbConnection conexion, string NroUsuario, int IDDeCliente)
        {
            Nuevo(conexion, IDDeCliente);
        }

       protected void Nuevo(System.Data.OleDb.OleDbConnection conexion, int IDdeCliente = 0)
       {
           DetallePedido = new Dictionary<string, PedidoItem>();
           mvarNroPedido = "";
           mvarF_Pedido = System.DateTime.Today;
           mvarConexion = conexion;

           if (IDdeCliente > 0)
               mvarIdCliente = IDdeCliente;
       }

       // FUNDAMENTAL PARA QUE TE DE LOS NOMBRES
       public PedidoItem Renglon(int Numero)
       {
           return DetallePedido["_" + Numero];
       }

        public System.Data.OleDb.OleDbConnection Conexion
        {
            set { mvarConexion = value; }
        }

        public byte NroImpresion
        {
            get { return mvarNroImpresion; }
            set { mvarNroImpresion = value; }
        }

        public System.DateTime F_Pedido
        {
            get { return mvarF_Pedido; }
        }

        public string NroPedido
        {
            get { return mvarNroPedido; }
        }

        private bool ValidarConexion()
        {
            return (mvarConexion != null);
        }

        public void Guardar(string Origen)
        {
            if (!(ValidarConexion()))
                return;

            System.Data.OleDb.OleDbDataReader rec = null;
            System.Data.OleDb.OleDbCommand cmd = new System.Data.OleDb.OleDbCommand();

            if (DetallePedido.Count < 1)
                return;

            if (Origen.ToUpper() == "VER")
            {
                Funciones.oleDbFunciones.ComandoIU(mvarConexion, "DELETE FROM tblPedido_Enc WHERE NroPedido='09999-99999999'");
                mvarNroPedido = "09999-99999999";
            }
            else
            {
                if (Global01.NroDocumentoAbierto.Trim().Length>0)
                {
                    Funciones.oleDbFunciones.ComandoIU(Global01.Conexion, "DELETE FROM tblPedido_Enc WHERE NroPedido='" + Global01.NroDocumentoAbierto + "'");
                    Global01.NroDocumentoAbierto = "";
                }

                rec = Funciones.oleDbFunciones.Comando(mvarConexion, "SELECT TOP 1 right(NroPedido,8) AS NroPedido FROM tblPedido_Enc WHERE left(NroPedido,5)=" + Global01.NroUsuario + " ORDER BY NroPedido DESC");
                if (!rec.HasRows)
                {
                    // ES UN CLIENTE
                    if (Global01.miSABOR == Global01.TiposDeCatalogo.Cliente)
                    {
                        mvarNroPedido = Global01.NroUsuario.Trim() + "-C0000001";
                    }
                    else
                    {
                        mvarNroPedido = Global01.NroUsuario.Trim() + "-00000001";
                    }
                }
                else
                {
                    rec.Read();
                    // ES UN CLIENTE
                    if (Global01.miSABOR == Global01.TiposDeCatalogo.Cliente)
                    {
                        mvarNroPedido = Global01.NroUsuario.Trim() + "-C" + (int.Parse(rec["NroPedido"].ToString().Substring(rec["NroPedido"].ToString().Length - 7)) + 1).ToString().PadLeft(7, '0');
                    }
                    else
                    {
                        mvarNroPedido = Global01.NroUsuario.Trim() + "-" + (int.Parse(rec["NroPedido"].ToString().Substring(rec["NroPedido"].ToString().Length - 8)) + 1).ToString().PadLeft(8,'0');
                    };
                }
                rec = null;

            }

            cmd.Parameters.Add("pNroPedido", System.Data.OleDb.OleDbType.VarChar, 14).Value = mvarNroPedido;
            cmd.Parameters.Add("pF_Pedido", System.Data.OleDb.OleDbType.Date).Value = mvarF_Pedido;
            cmd.Parameters.Add("pIdCliente", System.Data.OleDb.OleDbType.Integer).Value = mvarIdCliente;
            cmd.Parameters.Add("pNroImpresion", System.Data.OleDb.OleDbType.Integer).Value = 0;
            cmd.Parameters.Add("pObservaciones", System.Data.OleDb.OleDbType.VarChar, 128).Value = " "; // mvarObservaciones;
            cmd.Parameters.Add("pTransporte", System.Data.OleDb.OleDbType.VarChar, 128).Value = " "; // mvarTransporte;

            cmd.Connection = mvarConexion;
            cmd.CommandType = System.Data.CommandType.StoredProcedure;
            cmd.CommandText = "usp_Pedido_Enc_add";
            cmd.ExecuteNonQuery();

            cmd = null;

            // Ahora GUARDO los ITEMS de ESTE pedido

            int I = 0;
            double wTotal = 0;
            wTotal = 0;

            for (I = 1; I <= DetallePedido.Count; I++)
            {
                GuardarItem(Renglon(I).IDCatalogo, Renglon(I).cantidad, Renglon(I).Similar, Renglon(I).Bahia, Renglon(I).Oferta, Renglon(I).Deposito, Renglon(I).Precio, Renglon(I).Observaciones);
                wTotal = wTotal + (Renglon(I).cantidad * Renglon(I).Precio);
            }

            Global01.NroImprimir = mvarNroPedido;

            // Por POLITICA NUESTRA  -- >  SE BORRA
            if (Origen.ToUpper() == "VER")
            {
               Funciones.oleDbFunciones.ComandoIU(mvarConexion, "EXEC usp_Pedido_Transmicion_Upd '" + mvarNroPedido + "'");
            }
            else
            {
                auditoria.Auditor.instance.guardar(auditoria.Auditor.ObjetosAuditados.Pedido, auditoria.Auditor.AccionesAuditadas.EXITOSO, "cli:" + mvarIdCliente.ToString().Trim().PadLeft(6, '0') + " ped:" + mvarNroPedido + " tot:" + string.Format("{0:N2}", wTotal));
                //Nuevo();
                if (GuardoOK != null)
                {
                    GuardoOK(Global01.NroImprimir);
                }
            } 

        }

        public void ADDItem(string IDCatalogo, float Precio, int cantidad, bool Similar, byte Deposito, bool Oferta, string Observaciones)
        {
            //OJO chkEsOfertaBahia===Bahia

            PedidoItem mvarItem = new PedidoItem();

            mvarItem.IDCatalogo = IDCatalogo;
            mvarItem.Precio = Precio;
            mvarItem.cantidad = cantidad;
            mvarItem.Similar = Similar;
            mvarItem.Deposito = Deposito;
            mvarItem.Oferta = Oferta;
            mvarItem.Observaciones = Observaciones;

            mvarItem.Bahia = false;
            
            DetallePedido["_" + (DetallePedido.Count + 1).ToString()] = mvarItem;

            mvarItem = null;

        }

        private void GuardarItem(string IDCatalogo, int cantidad, bool Similar, bool Bahia, bool Oferta, byte Deposito, float Precio, string Observaciones)
        {
            if (!(ValidarConexion()))
                return;

            System.Data.OleDb.OleDbDataReader rec = null;
            System.Data.OleDb.OleDbCommand cmd = new System.Data.OleDb.OleDbCommand();

            cmd.Parameters.Add("pNroPedido", System.Data.OleDb.OleDbType.VarChar, 14).Value = mvarNroPedido;
            cmd.Parameters.Add("pIdCatalogo", System.Data.OleDb.OleDbType.VarChar, 38).Value = IDCatalogo;
            cmd.Parameters.Add("pCantidad", System.Data.OleDb.OleDbType.Integer).Value = cantidad;
            cmd.Parameters.Add("pSimilar", System.Data.OleDb.OleDbType.Boolean).Value = Similar;
            cmd.Parameters.Add("pOferta", System.Data.OleDb.OleDbType.Boolean).Value = Oferta;
            cmd.Parameters.Add("pBahia", System.Data.OleDb.OleDbType.Boolean).Value = Bahia;
            cmd.Parameters.Add("pDeposito", System.Data.OleDb.OleDbType.TinyInt).Value = Deposito;
            cmd.Parameters.Add("pPrecio", System.Data.OleDb.OleDbType.Single ).Value = Precio;
            cmd.Parameters.Add("pObservaciones", System.Data.OleDb.OleDbType.VarChar, 80).Value = Observaciones;

            cmd.Connection = mvarConexion;
            cmd.CommandType = System.Data.CommandType.StoredProcedure;
            cmd.CommandText = "usp_Pedido_Det_add";
            cmd.ExecuteNonQuery();

            rec = Funciones.oleDbFunciones.Comando(mvarConexion, "SELECT ID FROM CatalogoBAK WHERE ID='" + IDCatalogo + "'");
            if (!rec.HasRows)
            {
                Funciones.oleDbFunciones.ComandoIU(mvarConexion, "EXEC usp_AnexaItemCatalogoBak '" + IDCatalogo + "'");
            }
            else
            {
                Funciones.oleDbFunciones.ComandoIU(mvarConexion, "EXEC usp_CambiaCodigoCatalogoBak '" + IDCatalogo + "'");
            }
            rec = null;
            cmd = null;
     
        }


    }
}